[BugFix] check time type in DDL (#23473)

srlch opened the pull request:
## What type of PR is this:
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes:
<!--
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #23473

## Problem Summary(Required):
Problem:
Time data type is not support in storage layer, it just a data type of parameter or return value for some function.
It can not be specified by user in DDL

Solution:
check time type in DDL

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [x] 3.0
  - [x] 2.5
  - [ ] 2.4
  - [ ] 2.3

packy92 commented on fe/fe-core/src/main/java/com/starrocks/sql/parser/StarRocks.g4:
```diff
@@ -2365,7 +2365,6 @@ baseType
     | DATE
     | DATETIME
-    | TIME
     | CHAR typeParameter?
```
Now, cast(value as time) is supported. Maybe add check in DDL is better?
mysql> select cast(100 as time);
+-------------------+
| CAST(100 AS TIME) |
+-------------------+
| 00:01:00          |
+-------------------+
1 row in set (0.01 sec)

srlch replied:
ok

srlch committed:
[BugFix] check time type in DDL (#23473)

Problem:
Time data type is not support in storage layer, it just a data type of parameter or return value for some function.
It can not be specified by user in DDL

Solution:
check time type in DDL

Signed-off-by: srlch <linzichao@starrocks.com>

srlch commented:
> You need to check the time type when creating tables not in syntax file.

have fix it

packy92 commented on fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AlterTableStatementAnalyzer.java:
```diff
@@ -292,6 +292,10 @@ public Void visitAddColumnClause(AddColumnClause clause, ConnectContext context)
                 throw new SemanticException(PARSER_ERROR_MSG.invalidColumnDef(e.getMessage()), columnDef.getPos());
             }
 
+            if (columnDef.getType().isTime()) {
```
create table also need this check.

srlch replied:
create table have been checked in the current code. It just miss for DDL

wanpengfei-git commented:
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 4 / 4 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/AlterTableStatementAnalyzer.java|4|4|100.00%|[]|

imay commented on fe/fe-core/src/test/java/com/starrocks/sql/analyzer/AnalyzeAlterTableStatementTest.java:
```diff
@@ -141,4 +141,10 @@ public void testRollup() {
+    @Test
+    public void testAlterWithTimeType() {
+        analyzeFail("alter table t0 add column testcol TIME");
+        analyzeFail("alter table t0 modify column v0 TIME");
```
Why not add a `create table` test case to cover creating table with TIME type.

srlch replied:
create table test case have existed in the code. We just miss for DDL

wanpengfei-git commented:
@Mergifyio backport branch-3.0

wanpengfei-git commented:
@Mergifyio backport branch-2.5
